### PR TITLE
Revise UDP read thread implementation

### LINKED ISSUE #240: Revise UDP reading and copying workflow
* UDP reading on Win reads the msg to receiveBufferTmp, gets the lock, copies receiveBufferTmp to receiveBuffer and releases the lock in the read thread. Later, receiveBuffer is copied to the package or read buffer again, thus we have double copying of the mesg buffer.
* UDP reading on Linux gets the lock, reads the msg to msgInternal and releases the lock. Later, msgInternal is copied to the package or read buffer again, thus we have one memcpy less than compared to Win.

I believe the workflow on Win is fail-safer than the Linux one, but I doubt if it is really necessary. Would be good to have a consistent workflow on both implementations.

### PR DESCRIPTION
**Pull request by tbeu**
* Utilize WSAPoll on Windows and avoid memcpy between receiveBuffer and receiveBufferTmp
* Remove receiveBufferTmp from MDDUDPSocket_s on Windows
* Fix lock on nReceivedBytes on Linux
* Add missing break statement in switch on Linux

close #240

### CODE REVIEW

File: Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h [hunk 1]
```diff
@@ -296,21 +318,23 @@ void* MDD_udpReceivingThread(void * p_udp) {
...
+                        pthread_mutex_unlock(&(udp->messageMutex));
                         ModelicaFormatError("MDDUDPSocket.h: recvfrom(..) failed (%s)\n",
                                             strerror(errno));
                     }
-                    break;
+                    else {
+                        pthread_mutex_unlock(&(udp->messageMutex));
+                    }
                 }
+                break;
```
Comment (bernhard-thiele) [inline Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h]: OK, nicer to have the break outside the else branch (though the only possibility that the else branch is not called is that an error occured and `ModelicaFormatError` aborts the simulation anyway).

**Comment by bernhard-thiele** [comment]
Not sure what has been changed for the bullet "Fix lock on nReceivedBytes on Linux". Can't relate that bullet to any changes that I see in the diff.

File: Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h [hunk 2]
```diff
@@ -296,21 +318,23 @@ void* MDD_udpReceivingThread(void * p_udp) {
...
                     udp->nReceivedBytes =
-                        recvfrom(udp->sock,                   /* UDP socket */
+                        recvfrom(udp->sock,                    /* UDP socket */
                                  udp->msgInternal,             /* receive buffer */
                                  udp->messageLength,           /* max bytes to receive */
                                  0,                            /* no special flags */
                                  (struct sockaddr*) &(udp->sa),/* address of sender */
                                  &sa_len
                                 );
-                    pthread_mutex_unlock(&(udp->messageMutex));
```
[the remark this answers is not in the document]
Comment (bernhard-thiele) [reply Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h]: Ok, for that I would just put the unlock after the `if` branch. I don't see that there is a benefit to unlock the mutex before calling `ModelicaFormatError(..)`.

[the remark this answers is not in the document]
Comment (tbeu) [reply Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h]: > I don't see that there is a benefit to unlock the mutex before calling ModelicaFormatError(..).

You will get a reource leak then.

[the remark this answers is not in the document]
Comment (bernhard-thiele) [reply Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h]: Ok, I see your point now.

**Comment by tbeu** [comment]
Thanks. Will update ITI_MDD.dll later (after MQTT merge), since there was no change in the exported functions.

**Comment by bernhard-thiele** [comment]
Just did a quick test with OpenModelica on Windows. Unfortunately, the changes don't work with OM. I get errors similar to below.

```
In file included from Modelica_DeviceDrivers.Blocks.Examples.TestSerialPackager_UDP_includes.h:5:0,
                 from Modelica_DeviceDrivers.Blocks.Examples.TestSerialPackager_UDP_functions.c:6:
D:/ModelicaLibs/Modelica_DeviceDrivers/Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h: In function 'MDD_udpReceivingThread':
D:/ModelicaLibs/Modelica_DeviceDrivers/Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h:50:5: error: unknown type name 'WSAPOLLFD'
     WSAPOLLFD sock_poll = {0};
```

Maybe it is something simple like a missing header...

**Comment by bernhard-thiele** [comment]
@tbeu While trying tho fix the issue for OM, I realized that it is also not working with Dymola (using MSVC 2010). Seems that my test before did not use the MDD with the applied changes. I now get the following error: `WSAPoll failed with error: 10022`

**Comment by tbeu** [comment]
According to [MSDN](https://msdn.microsoft.com/de-de/library/windows/desktop/ms740094.aspx) WSAPOLLFD is defined in Winsock2.h, which is included for MSVC or MinGW: https://github.com/modelica/Modelica_DeviceDrivers/blob/2f88ba77353590b8f1e3d820fa05f8cac3ce52d4/Modelica_DeviceDrivers/Resources/Include/MDDUDPSocket.h#L28-L30

**Comment by bernhard-thiele** [comment]
Yes, but there seem to be issues with the header files that come with MinGW (at least for older versions)
https://virtuallyfun.com/wordpress/2017/02/11/wsapoll-mingw/

**Comment by tbeu** [comment]
Can you test the mentioned workaround?

**Comment by bernhard-thiele** [comment]
Yes, I did some definitions and it builds. However, I get the same error when executing as I now get with Dymola (`WSAPoll failed with error: 10022`)

**Comment by tbeu** [comment]
Hm, WSA error 10022 means WSAEINVAL = Invalid argument.

**Comment by tbeu** [comment]
OK, can reproduce.

**Comment by bernhard-thiele** [comment]
Instead of
`sock_poll.events = POLLIN | POLLHUP;`

One needs to do

```
    sock_poll.events = POLLIN;
    sock_poll.revents = POLLHUP;
```

**Comment by bernhard-thiele** [comment]
I can get things to simulate for OM if I put in following definitions

```
#if defined(__MINGW32__)
/* This is an (ugly) fix if MinGW headers miss definitions for WSAPoll */
#ifndef POLLIN
typedef struct pollfd {
    SOCKET fd;
    SHORT events;
    SHORT revents;
} WSAPOLLFD, *PWSAPOLLFD, FAR *LPWSAPOLLFD;
WINSOCK_API_LINKAGE int WSAAPI WSAPoll(LPWSAPOLLFD fdArray, ULONG fds, INT timeout);
#define POLLIN 768
#define POLLHUP 2
#endif /* POLLIN */
#endif /* defined(__MINGW32__) */
```

**Comment by tbeu** [comment]
Does it work if you set `struct pollfd` instead of `WSAPOLLFD`?

**Comment by tbeu** [comment]
> Instead of
`sock_poll.events = POLLIN | POLLHUP;`

> One needs to do
```
    sock_poll.events = POLLIN;
    sock_poll.revents = POLLHUP;
```

Confirmed.

**Comment by bernhard-thiele** [comment]
I suppose the `sock_poll.revents = POLLHUP;` should also be changed in the Linux part. I can't test Linux at the moment, but I will test it in the evening and just push a commit which fixes the issues

**Comment by bernhard-thiele** [comment]
After having a closer look the documentation(s) for pollfd, the only thing that I understand should be set in both Linux and Windows is

```
sock_poll.events = POLLIN;
```

The `revents` is the  output parameter, filled by the kernel. Seems like Linux simply ignores if `POLLHUP` is or'd into `events` while Windows throws an error.

**Comment by tbeu** [comment]
Thanks for the fix and tests.

**Pull request merged**